### ISSUE TITLE
Since 6.06, can't issue GET requests for HTTPS URLs through a proxy

### ISSUE THREAD

**Issue by sebasttj** [libwww-perl/libwww-perl#96]
Prior to version 6.06, when using a proxy for a request to an https url, LWP would issue a GET (or HEAD, or whatever) request to the proxy. E.g.,

```
https_proxy=http://localhost/ lwp-request https://example.com
```

was similar to

```
nc localhost 80 <<<$'GET https://example.com\n'
```

Since CONNECT support was added in 6.06, this feature seems to have been lost, with no way to enable the previous behavior. (If there _is_ some easy way to enable the old behavior, I'd love to hear it. It would make the rest of this message irrelevant!) The key place in the code is [where `$ssl_tunnel` gets set](https://github.com/libwww-perl/libwww-perl/blob/0e8a6ece4fed28871dd610e401a708de9980d050/lib/LWP/Protocol/http.pm#L151). If there were an option to force `$ssl_tunnel` to 0 rather than rely on that test, that would satisfy my needs. I'd be happy to submit a patch if that would be welcome.

"Why would anyone want that behavior rather than CONNECT?" you might ask. Sure. It allows a perl script to make a request that requires 2-way authentication (client certs) without giving that script access to the private key. If the [proxy knows the private key](https://httpd.apache.org/docs/current/mod/mod_ssl.html#sslproxymachinecertificatefile), the script does not need to know it. This is useful for scripts that run under mod_perl or CGI, which run in the context of an ostensibly unprivileged user.

There are a couple of ways to work around this that I've thought of. One is using a reverse proxy instead of a forward proxy. That works as long as the remote app's URL-space is compatible with the reverse proxy's; but it adds configuration complexity to the proxy server, and it requires that the perl application supports swapping out a portion of the URL it wants to connect to. Obviously not an ideal solution.

The other way that I've tried is by subclassing LWP::Protocol::http and URI::http (and *::https, if the proxy itself is https) to create a pseudo-protocol for the proxy (works just like http/https, but isn't _called_ http/https). It works great and is very little code, but I'm worried that it's brittle to future changes in LWP. It relies on the particular detail of the test that `$url->scheme eq 'https'`, and that changing what `$url->scheme` returns without changing `$url->as_string` doesn't break anything.

Thank you, and I'm looking forward to hearing your input.

**Comment by genio** [comment]
@sebasttj I'm not able to replicate your problem.  Please look through my setup below and point out where I'm differing from you as that would help me a great deal.

I setup [Squid](http://www.squid-cache.org/) to run on ```http://localhost:3128```. I then ran this small test script which returned the expected results.

```perl
#!/usr/bin/env perl

use strict;
use warnings;

use Data::Dumper::Concise;
use HTTP::Request ();
use LWP::UserAgent ();
use v5.20;

BEGIN {
    $ENV{PERL_LWP_ENV_PROXY} = 'http://localhost:3128';
    $ENV{https_proxy} = 'http://localhost:3128';
}

my $ua = LWP::UserAgent->new();
my $req = HTTP::Request->new('GET','https://google.com/');
my $response = $ua->request($req);
print $response->code ."\n";
say Dumper $response;
```

@noxxi Do you have any opinion on this? https://github.com/libwww-perl/libwww-perl/blob/0e8a6ece4fed28871dd610e401a708de9980d050/lib/LWP/Protocol/http.pm#L151

**Comment by noxxi** [comment]
If I understand @sebasttj  correctly he wants a way to have a plain text connection to the proxy configured by https_proxy, tell the proxy to retrieve a `https://` URL and then send the response back to the client as plain text. This is the behavior LWP had between 6.00 and 6.05 and which was fixed in 6.06 because it is not the expected behavior when talking to a HTTP proxy and will in fact not work with most proxies.

In my opinion the behavior @sebasttj wants is not covered by any standard, not implemented by major browsers and not implemented by major proxies. It is also not the behavior one usually associates with a HTTP proxy and the way it worked from 6.00 to 6.05 was simply a bug which was fixed in 6.06. I'm not saying that this is not a valid use case but I would not cover this  with the normal HTTP proxy behavior as implemented in 6.06+ because it simply is not the normal HTTP proxy behavior. If @sebasttj  needs this use case then I recommend that he writes a custom `LWP::Protocol::...` connector for it instead.

**Comment by noxxi** [comment]
> @noxxi Do you have any opinion on this? https://github.com/libwww-perl/libwww-perl/blob/0e8a6ece4fed28871dd610e401a708de9980d050/lib/LWP/Protocol/http.pm#L151

I'm not sure what kind of opinion I should have. But it was me who wrote this code, see cb80c2dd.

**Comment by genio** [comment]
@noxxi Thanks for the description and I think you are right about the scenario @sebasttj is describing.

@sebasttj Given the explanation above, it seems like it would be a bug to re-introduce the functionality you're describing. As was suggested, you could write your own ```LWP::Protocol::``` that borrows from ```LWP::Protocol::http``` and the older way it worked with proxy servers.

I'm going to close this out for now under the assumption that the current functionality of LWP is as it should be.  Please don't hesitate to talk it through with us on IRC or reopen the issue if you would like to discuss it more.

Thanks,
Chase

**Comment by sebasttj** [comment]
@noxxi, Yes, that's exactly what I want. And it is supported by both squid and Apache httpd/mod_proxy (and the proxy's use of client certificates is supported in mod_ssl via the [SSLMachineCertificateFile directive](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslproxymachinecertificatefile), which allows for 2-way PKI auth from both forward and reverse proxies).

You can try it yourself. This works just fine for me with squid in a default configuration on CentOS 7:

    (echo $'GET https://google.com HTTP/1.0\n'; sleep 2) | nc localhost 3128 

I've looked through the standards, and I don't see what would make that request non-compliant. Though I admit I'm not an expert in that arena, so I very well could be wrong.

To summarize:

* LWP used to support this.
* It is currently supported by at least two major proxies (squid and mod_proxy).
* It has a reasonable use-case (client-agnostic two-way PKI auth, though that might only work with mod_proxy, I don't know).

But if you don't want to support it, then that's that. As I said I already have two work-arounds, so this isn't a showstopper or anything. Thanks for your time.

**Comment by thoke** [comment]
I looked into this behavior a long time ago and my opinion is that squid
and mod_proxy are behaving in an incorrect (non-secure even) way. I noticed
the different behavior as we needed to start working with Microsoft Threat
Management Gateway which does NOT allow the direct GET/HEAD functionality
and requires the CONNECT mechanism. For me, the authentication part is what
got really tricky. Now that I'm going through module updates, I'm afraid
our authentication is going to break (but that's another thread when I get
there).

-Tim

**Comment by noxxi** [comment]
> You can try it yourself. This works just fine for me with squid in a default configuration on CentOS 7:

I've tried. Freshly installed squid 3.5.12 on Ubuntu in default setup gives in this case an error: `Squid does not support all request methods for all access protocols`. Same result with a different system and squid 3.1.19.  But you are right in that squid 3.5.20 on CentOS7 allows such requests. It must be something compiled in because the the config on Ubuntu does not allow the requests. Strange.
Anyway http://wiki.squid-cache.org/Features/HTTPS clearly describes what is expected from a client which wants to use squid as a HTTPS proxy: do a CONNECT request or talk directly TLS (in case of a transparent proxy).

> LWP used to support this.

It was not support in LWP < 6, i.e. in connection with Crypt::SSLeay. While it was kind of "supported" in LWP 6.0..6.05 it actually was a bug since it did not implement this behavior additionally to the typical use of a HTTPS proxy but instead - in other words: it implemented HTTPS proxy support the wrong way and only the wrong way. And users rightfully complained about it and tried workarounds like LWP::Protocol::connect or Net::SSLGlue::LWP.

> It is currently supported by at least two major proxies (squid and mod_proxy).

I don't know why these proxies support it (or don't, see above) but no browser supports this.

> It has a reasonable use-case (client-agnostic two-way PKI auth, though that might only work with mod_proxy, I don't know).

I don't doubt that there is a use case. But again, this is not the behavior one normally expects from a HTTP proxy (and probably the reason no browser supports it) but instead one expects end-to-end security with HTTPS even when using a proxy. That's why I recommend to implement this use case as a special LWP::Protocol::* module.

**Comment by sebasttj** [comment]
Ok, that all sounds perfectly reasonable. Thanks for taking the time to talk about it with me.

I did implement my own protocol module, and it seems to work fine.
